pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 					ident.StoreID, desc)
 			}
 
-			initM[storage.FullReplicaID{
-				RangeID:   desc.RangeID,
-				ReplicaID: repDesc.ReplicaID,
-			}] = &desc
+			if err := s.setDesc(desc.RangeID, desc); err != nil {
+				return err
+			}
+			s.setDescReplicaID(desc.RangeID, repDesc.ReplicaID)
 			return nil
 		}); err != nil {
-		return nil, err
+		return LoadedReplicas{}, err
 	}
 
-	allM, err := loadFullReplicaIDsFromDisk(ctx, eng)
-	if err != nil {
-		return nil, err
+	// INVARIANT: all replicas have a persisted full ReplicaID (i.e. a "ReplicaID from disk").
+	//
+	// This invariant is true for replicas created in 22.1. Without further action, it
+	// would be violated for clusters that originated before 22.1. In this method, we
+	// backfill the ReplicaID (for initialized replicas) and we remove uninitialized
+	// replicas (see below for rationale).
```

Comment:
Removing all uninitialized replicas? Or the ones that don't have a `RaftReplicaID`? Worth clarifying here and in the commit description.

Reply:
Done.

Reply:
(in comment, commit, PR)